// @luma.gl/core/dist/es5/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "assert", {
  enumerable: true,
  get: function get() {
    return _webgl.assert;
  }
});
Object.defineProperty(exports, "log", {
  enumerable: true,
  get: function get() {
    return _webgl.log;
  }
});
Object.defineProperty(exports, "uid", {
  enumerable: true,
  get: function get() {
    return _webgl.uid;
  }
});
Object.defineProperty(exports, "isObjectEmpty", {
  enumerable: true,
  get: function get() {
    return _webgl.isObjectEmpty;
  }
});
Object.defineProperty(exports, "self", {
  enumerable: true,
  get: function get() {
    return _webgl.self;
  }
});
Object.defineProperty(exports, "window", {
  enumerable: true,
  get: function get() {
    return _webgl.window;
  }
});
Object.defineProperty(exports, "global", {
  enumerable: true,
  get: function get() {
    return _webgl.global;
  }
});
Object.defineProperty(exports, "document", {
  enumerable: true,
  get: function get() {
    return _webgl.document;
  }
});

var _webgl = require("@luma.gl/webgl");
