// @superset-ui/preset-chart-xy/lib/components/legend/DefaultLegendItem.js
"use strict";

exports.__esModule = true;
exports.default = DefaultLegendItem;

var _react = _interopRequireDefault(require("react"));

var _legend = require("@vx/legend");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MARK_SIZE = 8;
const MARK_STYLE = {
  display: 'inline-block'
};

function DefaultLegendItem({
  group,
  item,
  MarkRenderer,
  LabelRenderer
}) {
  var _ref, _ref2, _item$output$color, _item$output$stroke;

  return (0, _react2.jsx)(_legend.LegendItem, {
    key: `legend-item-${group.field}-${item.input}`,
    margin: "0 5px"
  }, typeof MarkRenderer === 'undefined' ? (0, _react2.jsx)("svg", {
    width: MARK_SIZE,
    height: MARK_SIZE,
    style: MARK_STYLE
  }, (0, _react2.jsx)("circle", {
    fill: // @ts-ignore
    (_ref = (_ref2 = (_item$output$color = item.output.color) != null ? _item$output$color : // @ts-ignore
    item.output.fill) != null ? _ref2 : // @ts-ignore
    item.output.stroke) != null ? _ref : '#ccc',
    stroke: // @ts-ignore
    (_item$output$stroke = item.output.stroke) != null ? _item$output$stroke : 'none',
    r: MARK_SIZE / 2,
    cx: MARK_SIZE / 2,
    cy: MARK_SIZE / 2
  })) : (0, _react2.jsx)(MarkRenderer, {
    group: group,
    item: item
  }), typeof LabelRenderer === 'undefined' ? (0, _react2.jsx)(_legend.LegendLabel, {
    align: "left",
    margin: "0 0 0 4px"
  }, item.input) : (0, _react2.jsx)(LabelRenderer, {
    group: group,
    item: item
  }));
}
