// @luma.gl/webgl/dist/es5/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "assert", {
  enumerable: true,
  get: function get() {
    return _assert["default"];
  }
});
Object.defineProperty(exports, "log", {
  enumerable: true,
  get: function get() {
    return _log["default"];
  }
});
Object.defineProperty(exports, "isBrowser", {
  enumerable: true,
  get: function get() {
    return _isBrowser["default"];
  }
});
Object.defineProperty(exports, "isOldIE", {
  enumerable: true,
  get: function get() {
    return _isOldIe["default"];
  }
});
Object.defineProperty(exports, "uid", {
  enumerable: true,
  get: function get() {
    return _utils.uid;
  }
});
Object.defineProperty(exports, "isPowerOfTwo", {
  enumerable: true,
  get: function get() {
    return _utils.isPowerOfTwo;
  }
});
Object.defineProperty(exports, "isObjectEmpty", {
  enumerable: true,
  get: function get() {
    return _utils.isObjectEmpty;
  }
});
Object.defineProperty(exports, "formatValue", {
  enumerable: true,
  get: function get() {
    return _formatValue.formatValue;
  }
});
Object.defineProperty(exports, "stubRemovedMethods", {
  enumerable: true,
  get: function get() {
    return _stubMethods.stubRemovedMethods;
  }
});
Object.defineProperty(exports, "checkProps", {
  enumerable: true,
  get: function get() {
    return _checkProps.checkProps;
  }
});
Object.defineProperty(exports, "cssToDevicePixels", {
  enumerable: true,
  get: function get() {
    return _devicePixels.cssToDevicePixels;
  }
});
Object.defineProperty(exports, "cssToDeviceRatio", {
  enumerable: true,
  get: function get() {
    return _devicePixels.cssToDeviceRatio;
  }
});
Object.defineProperty(exports, "getDevicePixelRatio", {
  enumerable: true,
  get: function get() {
    return _devicePixels.getDevicePixelRatio;
  }
});

var _assert = _interopRequireDefault(require("./assert"));

var _log = _interopRequireDefault(require("./log"));

var _isBrowser = _interopRequireDefault(require("./is-browser"));

var _isOldIe = _interopRequireDefault(require("./is-old-ie"));

var _utils = require("./utils");

var _formatValue = require("./format-value");

var _stubMethods = require("./stub-methods");

var _checkProps = require("./check-props");

var _devicePixels = require("./device-pixels");
