// @luma.gl/webgl-state-tracker/dist/es5/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getParameter", {
  enumerable: true,
  get: function get() {
    return _unifiedParameterApi.getParameter;
  }
});
Object.defineProperty(exports, "getParameters", {
  enumerable: true,
  get: function get() {
    return _unifiedParameterApi.getParameters;
  }
});
Object.defineProperty(exports, "setParameter", {
  enumerable: true,
  get: function get() {
    return _unifiedParameterApi.setParameter;
  }
});
Object.defineProperty(exports, "resetParameters", {
  enumerable: true,
  get: function get() {
    return _unifiedParameterApi.resetParameters;
  }
});
Object.defineProperty(exports, "getModifiedParameters", {
  enumerable: true,
  get: function get() {
    return _unifiedParameterApi.getModifiedParameters;
  }
});
Object.defineProperty(exports, "setParameters", {
  enumerable: true,
  get: function get() {
    return _setParameters.setParameters;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _trackContextState["default"];
  }
});
Object.defineProperty(exports, "trackContextState", {
  enumerable: true,
  get: function get() {
    return _trackContextState["default"];
  }
});
Object.defineProperty(exports, "pushContextState", {
  enumerable: true,
  get: function get() {
    return _trackContextState.pushContextState;
  }
});
Object.defineProperty(exports, "popContextState", {
  enumerable: true,
  get: function get() {
    return _trackContextState.popContextState;
  }
});
Object.defineProperty(exports, "withParameters", {
  enumerable: true,
  get: function get() {
    return _withParameters.withParameters;
  }
});

var _unifiedParameterApi = require("./unified-parameter-api/unified-parameter-api");

var _setParameters = require("./unified-parameter-api/set-parameters");

var _trackContextState = _interopRequireWildcard(require("./state-tracking/track-context-state"));

var _withParameters = require("./state-tracking/with-parameters");
