// @luma.gl/core/dist/es6/multipass/shader-module-pass.js
import Pass from './pass';
import CompositePass from './composite-pass';
import ClipSpace from '../lib/clip-space';
import { normalizeShaderModule } from '@luma.gl/shadertools';

class ShaderModuleSinglePass extends Pass {
  constructor(gl, props = {}) {
    super(gl, Object.assign({
      swap: true
    }, props));
  }

  _renderPass({
    inputBuffer,
    swapBuffers
  }) {
    this.props.model.setUniforms(this.props);
    this.props.model.draw({
      uniforms: {
        texture: inputBuffer,
        texSize: [inputBuffer.width, inputBuffer.height]
      },
      parameters: {
        depthWrite: false,
        depthTest: false
      }
    });
  }

}

export default class ShaderModulePass extends CompositePass {
  constructor(gl, module, props = {}) {
    const id = "".concat(module.name, "-pass");
    normalizeShaderModule(module);
    const passes = normalizePasses(gl, module, id, props);
    super(gl, Object.assign({
      id,
      passes
    }, props));
    this.module = module;
  }

  _renderPass({
    inputBuffer,
    swapBuffers
  }) {
    let first = true;

    for (const pass of this.props.passes) {
      if (!first) {
        swapBuffers();
      }

      first = false;
      const {
        uniforms,
        model
      } = pass.props;

      if (uniforms) {
        model.setUniforms(uniforms);
      }

      model.draw({
        uniforms: {
          texture: inputBuffer,
          texSize: [inputBuffer.width, inputBuffer.height]
        },
        parameters: {
          depthWrite: false,
          depthTest: false
        }
      });
    }
  }

}

function normalizePasses(gl, module, id, props) {
  if (module.filter || module.sampler) {
    const fs = getFragmentShaderForRenderPass(module);
    const pass = new ShaderModuleSinglePass(gl, {
      id,
      model: getModel(gl, module, fs, id, props),
      uniforms: null
    });
    return [pass];
  }

  const passes = module.passes || [];
  return passes.map(pass => {
    const fs = getFragmentShaderForRenderPass(module, pass);
    const idn = "".concat(id, "-").concat(passes.length + 1);
    return new ShaderModuleSinglePass(gl, Object.assign({
      id: idn,
      model: getModel(gl, module, fs, idn, props),
      uniforms: pass.uniforms
    }, props));
  });
}

function getModel(gl, module, fs, id, props) {
  const model = new ClipSpace(gl, {
    id,
    fs,
    modules: [module]
  });
  const uniforms = Object.assign(module.getUniforms(), module.getUniforms(props));
  model.setUniforms(uniforms);
  return model;
}

const FILTER_FS_TEMPLATE = func => "uniform sampler2D texture;\nuniform vec2 texSize;\n\nvarying vec2 position;\nvarying vec2 coordinate;\nvarying vec2 uv;\n\nvoid main() {\n  vec2 texCoord = coordinate;\n\n  gl_FragColor = texture2D(texture, texCoord);\n  gl_FragColor = ".concat(func, "(gl_FragColor, texSize, texCoord);\n}\n");

const SAMPLER_FS_TEMPLATE = func => "uniform sampler2D texture;\nuniform vec2 texSize;\n\nvarying vec2 position;\nvarying vec2 coordinate;\nvarying vec2 uv;\n\nvoid main() {\n  vec2 texCoord = coordinate;\n\n  gl_FragColor = ".concat(func, "(texture, texSize, texCoord);\n}\n");

function getFragmentShaderForRenderPass(module, pass = module) {
  if (pass.filter) {
    const func = typeof pass.filter === 'string' ? pass.filter : "".concat(module.name, "_filterColor");
    return FILTER_FS_TEMPLATE(func);
  }

  if (pass.sampler) {
    const func = typeof pass.sampler === 'string' ? pass.sampler : "".concat(module.name, "_sampleColor");
    return SAMPLER_FS_TEMPLATE(func);
  }

  return null;
}
