// @superset-ui/preset-chart-xy/lib/components/Line/DefaultTooltipRenderer.js
"use strict";

exports.__esModule = true;
exports.default = DefaultTooltipRenderer;

var _react = _interopRequireDefault(require("react"));

var _core = require("@superset-ui/core");

var _theme = require("@data-ui/theme");

var _react2 = require("@emotion/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MARK_STYLE = {
  marginRight: 4
};

function DefaultTooltipRenderer({
  allSeries,
  datum,
  encoder,
  series = {},
  theme = _theme.chartTheme
}) {
  return (0, _react2.jsx)(_core.TooltipFrame, null, (0, _react2.jsx)(_react.default.Fragment, null, (0, _react2.jsx)("div", {
    style: {
      fontFamily: theme.labelStyles.fontFamily
    }
  }, (0, _react2.jsx)("strong", null, encoder.channels.x.formatValue(datum.x))), (0, _react2.jsx)("br", null), series && (0, _react2.jsx)(_core.TooltipTable, {
    data: allSeries.filter(({
      key
    }) => series[key]).concat().sort((a, b) => series[b.key].y - series[a.key].y).map(({
      key,
      stroke,
      strokeDasharray,
      strokeWidth
    }) => ({
      key,
      keyColumn: (0, _react2.jsx)(_react.default.Fragment, null, (0, _react2.jsx)("svg", {
        width: "12",
        height: "8",
        style: MARK_STYLE
      }, (0, _react2.jsx)("line", {
        x2: "12",
        y1: "3",
        y2: "3",
        stroke: stroke,
        strokeWidth: strokeWidth,
        strokeDasharray: strokeDasharray
      })), series[key] === datum ? (0, _react2.jsx)("b", null, key) : key),
      valueColumn: encoder.channels.y.formatValue(series[key].y)
    }))
  })));
}
