// @loaders.gl/images/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ImageLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoader["default"];
  }
});
Object.defineProperty(exports, "HTMLImageLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoader.HTMLImageLoader;
  }
});
Object.defineProperty(exports, "ImageBitmapLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoader.ImageBitmapLoader;
  }
});
Object.defineProperty(exports, "ImageWriter", {
  enumerable: true,
  get: function get() {
    return _imageWriter["default"];
  }
});
Object.defineProperty(exports, "loadImage", {
  enumerable: true,
  get: function get() {
    return _parseImage.loadImage;
  }
});
Object.defineProperty(exports, "isImage", {
  enumerable: true,
  get: function get() {
    return _getImageMetadata.isImage;
  }
});
Object.defineProperty(exports, "getImageMetadata", {
  enumerable: true,
  get: function get() {
    return _getImageMetadata.getImageMetadata;
  }
});
Object.defineProperty(exports, "getImageMIMEType", {
  enumerable: true,
  get: function get() {
    return _getImageMetadata.getImageMIMEType;
  }
});
Object.defineProperty(exports, "getImageSize", {
  enumerable: true,
  get: function get() {
    return _getImageMetadata.getImageSize;
  }
});
Object.defineProperty(exports, "decodeImage", {
  enumerable: true,
  get: function get() {
    return _imageUtilsBrowser.decodeImage;
  }
});

var _imageLoader = _interopRequireWildcard(require("./image-loader"));

var _imageWriter = _interopRequireDefault(require("./image-writer"));

var _parseImage = require("./lib/parse-image");

var _getImageMetadata = require("./lib/get-image-metadata");

var _imageUtilsBrowser = require("./lib/image-utils-browser");
