// @superset-ui/preset-chart-xy/esm/components/Line/DefaultTooltipRenderer.js
import React from 'react';
import { TooltipFrame, TooltipTable } from '@superset-ui/core';
import { chartTheme } from '@data-ui/theme';
import { jsx as ___EmotionJSX } from "@emotion/react";
const MARK_STYLE = {
  marginRight: 4
};
export default function DefaultTooltipRenderer({
  allSeries,
  datum,
  encoder,
  series = {},
  theme = chartTheme
}) {
  return ___EmotionJSX(TooltipFrame, null, ___EmotionJSX(React.Fragment, null, ___EmotionJSX("div", {
    style: {
      fontFamily: theme.labelStyles.fontFamily
    }
  }, ___EmotionJSX("strong", null, encoder.channels.x.formatValue(datum.x))), ___EmotionJSX("br", null), series && ___EmotionJSX(TooltipTable, {
    data: allSeries.filter(({
      key
    }) => series[key]).concat().sort((a, b) => series[b.key].y - series[a.key].y).map(({
      key,
      stroke,
      strokeDasharray,
      strokeWidth
    }) => ({
      key,
      keyColumn: ___EmotionJSX(React.Fragment, null, ___EmotionJSX("svg", {
        width: "12",
        height: "8",
        style: MARK_STYLE
      }, ___EmotionJSX("line", {
        x2: "12",
        y1: "3",
        y2: "3",
        stroke: stroke,
        strokeWidth: strokeWidth,
        strokeDasharray: strokeDasharray
      })), series[key] === datum ? ___EmotionJSX("b", null, key) : key),
      valueColumn: encoder.channels.y.formatValue(series[key].y)
    }))
  })));
}
