// storybook-addon-paddings/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withPaddings = void 0;

var _addons = require("@storybook/addons");

var _constants = require("./constants");

var _helpers = require("./helpers");

var state = {
  initialized: false
};

var setStyle = function setStyle() {
  var padding = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _constants.DEFAULT_PADDING;
  document.body.style.margin = '0';
  document.body.style.padding = padding;
  document.body.style.transition = 'padding .3s';
};

var setPadding = function setPadding(padding) {
  state.padding = padding;
  setStyle(padding);
};

var render = function render(settings, storyFn) {
  var _settings$parameters = settings.parameters,
      options = _settings$parameters === void 0 ? {} : _settings$parameters;

  var channel = _addons.addons.getChannel();

  if (!state.initialized) {
    channel.on(_constants.EVENTS.UPDATE, setPadding);
    state.initialized = true;
  }

  var currentValue = (0, _helpers.isEnabled)(options) ? state.padding : _constants.DEFAULT_PADDING;
  setStyle((0, _helpers.getSelectedPadding)(options, currentValue));
  return storyFn();
};

var wrapper = function wrapper(getStory, context, settings) {
  return render(settings, function () {
    return getStory(context);
  });
};

var withPaddings = (0, _addons.makeDecorator)({
  name: 'withPaddings',
  parameterName: _constants.PARAM_KEY,
  skipIfNoParametersOrOptions: true,
  wrapper: wrapper
});
exports.withPaddings = withPaddings;
