// @superset-ui/plugin-chart-pivot-table/lib/plugin/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _core = require("@superset-ui/core");

var _chartControls = require("@superset-ui/chart-controls");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const config = {
  controlPanelSections: [{ ..._chartControls.sections.legacyTimeseriesTime,
    expanded: false
  }, {
    label: (0, _core.t)('Query'),
    expanded: true,
    controlSetRows: [[{
      name: 'groupbyRows',
      config: { ..._chartControls.sharedControls.groupby,
        label: (0, _core.t)('Rows'),
        description: (0, _core.t)('Columns to group by on the rows')
      }
    }], [{
      name: 'groupbyColumns',
      config: { ..._chartControls.sharedControls.groupby,
        label: (0, _core.t)('Columns'),
        description: (0, _core.t)('Columns to group by on the columns')
      }
    }], [{
      name: 'metrics',
      config: { ..._chartControls.sharedControls.metrics,
        validators: [_core.validateNonEmpty]
      }
    }], ['adhoc_filters'], [{
      name: 'row_limit',
      config: { ..._chartControls.sharedControls.row_limit
      }
    }]]
  }, {
    label: (0, _core.t)('Options'),
    expanded: true,
    tabOverride: 'data',
    controlSetRows: [[{
      name: 'aggregateFunction',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Aggregation function'),
        clearable: false,
        choices: (0, _chartControls.formatSelectOptions)(['Count', 'Count Unique Values', 'List Unique Values', 'Sum', 'Average', 'Median', 'Sample Variance', 'Sample Standard Deviation', 'Minimum', 'Maximum', 'First', 'Last', 'Sum as Fraction of Total', 'Sum as Fraction of Rows', 'Sum as Fraction of Columns', 'Count as Fraction of Total', 'Count as Fraction of Rows', 'Count as Fraction of Columns']),
        default: 'Sum',
        description: (0, _core.t)('Aggregate function to apply when pivoting and computing the total rows and columns'),
        renderTrigger: true
      }
    }], [{
      name: 'transposePivot',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Transpose pivot'),
        default: false,
        description: (0, _core.t)('Swap rows and columns'),
        renderTrigger: true
      }
    }]]
  }, {
    label: (0, _core.t)('Options'),
    expanded: true,
    controlSetRows: [[{
      name: 'valueFormat',
      config: { ..._chartControls.sharedControls.y_axis_format,
        label: (0, _core.t)('Value format')
      }
    }], [{
      name: 'tableRenderer',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Pivot table type'),
        default: 'Table With Subtotal',
        choices: [// [value, label]
        ['Table With Subtotal', (0, _core.t)('Table')], ['Table With Subtotal Heatmap', (0, _core.t)('Table Heatmap')], ['Table With Subtotal Col Heatmap', (0, _core.t)('Table Col Heatmap')], ['Table With Subtotal Row Heatmap', (0, _core.t)('Table Row Heatmap')], ['Table With Subtotal Barchart', (0, _core.t)('Table Barchart')], ['Table With Subtotal Col Barchart', (0, _core.t)('Table Col Barchart')], ['Table With Subtotal Row Barchart', (0, _core.t)('Table Row Barchart')]],
        renderTrigger: true,
        description: (0, _core.t)('The type of pivot table visualization')
      }
    }], [{
      name: 'rowOrder',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Rows sort by'),
        default: 'key_a_to_z',
        choices: [// [value, label]
        ['key_a_to_z', (0, _core.t)('key a-z')], ['key_z_to_a', (0, _core.t)('key z-a')], ['value_a_to_z', (0, _core.t)('value ascending')], ['value_z_to_a', (0, _core.t)('value descending')]],
        renderTrigger: true,
        description: (0, _core.t)('Order of rows')
      }
    }, {
      name: 'colOrder',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Cols sort by'),
        default: 'key_a_to_z',
        choices: [// [value, label]
        ['key_a_to_z', (0, _core.t)('key a-z')], ['key_z_to_a', (0, _core.t)('key z-a')], ['value_a_to_z', (0, _core.t)('value ascending')], ['value_z_to_a', (0, _core.t)('value descending')]],
        renderTrigger: true,
        description: (0, _core.t)('Order of columns')
      }
    }], [{
      name: 'rowSubtotalPosition',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Rows subtotals position'),
        default: false,
        choices: [// [value, label]
        [true, (0, _core.t)('Top')], [false, (0, _core.t)('Bottom')]],
        renderTrigger: true,
        description: (0, _core.t)('Position of row level subtotals')
      }
    }, {
      name: 'colSubtotalPosition',
      config: {
        type: 'SelectControl',
        label: (0, _core.t)('Cols subtotals position'),
        default: false,
        choices: [// [value, label]
        [true, (0, _core.t)('Left')], [false, (0, _core.t)('Right')]],
        renderTrigger: true,
        description: (0, _core.t)('Position of column level subtotals')
      }
    }], [{
      name: 'rowTotals',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Show rows totals'),
        default: true,
        renderTrigger: true,
        description: (0, _core.t)('Display row level totals')
      }
    }, {
      name: 'colTotals',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Show cols totals'),
        default: true,
        renderTrigger: true,
        description: (0, _core.t)('Display column level totals')
      }
    }], (0, _core.isFeatureEnabled)(_core.FeatureFlag.DASHBOARD_CROSS_FILTERS) ? [{
      name: 'emitFilter',
      config: {
        type: 'CheckboxControl',
        label: (0, _core.t)('Enable emitting filters'),
        renderTrigger: true,
        default: false,
        description: (0, _core.t)('Whether to apply filter to dashboards when table cells are clicked')
      }
    }] : []]
  }]
};
var _default = config;
exports.default = _default;
