// @superset-ui/preset-chart-xy/esm/components/legend/DefaultLegend.js
import React, { PureComponent } from 'react';
import DefaultLegendGroup from './DefaultLegendGroup';
import { jsx as ___EmotionJSX } from "@emotion/react";
const LEGEND_CONTAINER_STYLE = {
  display: 'flex',
  flexBasis: 'auto',
  flexGrow: 1,
  flexShrink: 1,
  maxHeight: 100,
  overflowY: 'auto',
  position: 'relative'
};
export default class DefaultLegend extends PureComponent {
  render() {
    const {
      groups,
      LegendGroupRenderer = DefaultLegendGroup,
      LegendItemRenderer,
      LegendItemMarkRenderer,
      LegendItemLabelRenderer,
      style
    } = this.props;
    const combinedStyle = typeof style === 'undefined' ? LEGEND_CONTAINER_STYLE : { ...LEGEND_CONTAINER_STYLE,
      ...style
    };
    return ___EmotionJSX("div", {
      style: combinedStyle
    }, groups.filter(group => 'items' in group && group.items.length > 0).map(group => ___EmotionJSX(LegendGroupRenderer, {
      key: group.field,
      group: group,
      ItemRenderer: LegendItemRenderer,
      ItemMarkRenderer: LegendItemMarkRenderer,
      ItemLabelRenderer: LegendItemLabelRenderer
    })));
  }

}
