// @superset-ui/preset-chart-xy/esm/components/legend/DefaultLegendItem.js
import React from 'react';
import { LegendItem, LegendLabel } from '@vx/legend';
import { jsx as ___EmotionJSX } from "@emotion/react";
const MARK_SIZE = 8;
const MARK_STYLE = {
  display: 'inline-block'
};
export default function DefaultLegendItem({
  group,
  item,
  MarkRenderer,
  LabelRenderer
}) {
  var _ref, _ref2, _item$output$color, _item$output$stroke;

  return ___EmotionJSX(LegendItem, {
    key: `legend-item-${group.field}-${item.input}`,
    margin: "0 5px"
  }, typeof MarkRenderer === 'undefined' ? ___EmotionJSX("svg", {
    width: MARK_SIZE,
    height: MARK_SIZE,
    style: MARK_STYLE
  }, ___EmotionJSX("circle", {
    fill: // @ts-ignore
    (_ref = (_ref2 = (_item$output$color = item.output.color) != null ? _item$output$color : // @ts-ignore
    item.output.fill) != null ? _ref2 : // @ts-ignore
    item.output.stroke) != null ? _ref : '#ccc',
    stroke: // @ts-ignore
    (_item$output$stroke = item.output.stroke) != null ? _item$output$stroke : 'none',
    r: MARK_SIZE / 2,
    cx: MARK_SIZE / 2,
    cy: MARK_SIZE / 2
  })) : ___EmotionJSX(MarkRenderer, {
    group: group,
    item: item
  }), typeof LabelRenderer === 'undefined' ? ___EmotionJSX(LegendLabel, {
    align: "left",
    margin: "0 0 0 4px"
  }, item.input) : ___EmotionJSX(LabelRenderer, {
    group: group,
    item: item
  }));
}
