// @superset-ui/preset-chart-xy/esm/Line/legacy/transformProps.js
import { flatMap } from 'lodash';
export default function transformProps(chartProps) {
  const {
    width,
    height,
    formData,
    queriesData
  } = chartProps;
  const {
    colorScheme,
    xAxisLabel,
    xAxisFormat,
    yAxisLabel,
    yAxisFormat
  } = formData;
  const data = queriesData[0].data;
  return {
    data: flatMap(data.map(row => row.values.map(v => ({ ...v,
      name: row.key[0]
    })))),
    width,
    height,
    encoding: {
      x: {
        field: 'x',
        type: 'temporal',
        format: xAxisFormat,
        scale: {
          type: 'time'
        },
        axis: {
          orient: 'bottom',
          title: xAxisLabel
        }
      },
      y: {
        field: 'y',
        type: 'quantitative',
        format: yAxisFormat,
        scale: {
          type: 'linear'
        },
        axis: {
          orient: 'left',
          title: yAxisLabel
        }
      },
      stroke: {
        field: 'name',
        type: 'nominal',
        scale: {
          scheme: colorScheme
        },
        legend: true
      }
    }
  };
}
